examples/components/MNIST_VERTICAL/traininsilo/host.py

```diff
@@ -0,0 +1,446 @@
+        self.loss_ = nn.CrossEntropyLoss()
+
+        self.train_dataset_, self.test_dataset_ = self.load_dataset(
+            train_data_dir, test_data_dir
+        )
+        self.train_loader_ = DataLoader(
+            self.train_dataset_, batch_size=batch_size, shuffle=False
+        )
+        self.test_loader_ = DataLoader(
+            self.test_dataset_, batch_size=batch_size, shuffle=False
+        )
+
+        self.model_.to(self.device_)
+        self._model_path = model_path
+
+        self.optimizer_ = SGD(self.model_.parameters(), lr=self._lr, momentum=0.9)
+
+    def load_dataset(self, train_data_dir, test_data_dir):
+        """Load dataset from {train_data_dir} and {test_data_dir}
+
+        Args:
+            train_data_dir(str, optional): Training data directory path
+            test_data_dir(str, optional): Testing data directory path
+        """
+        logger.info(f"Train data dir: {train_data_dir}, Test data dir: {test_data_dir}")
+        train_dataset = TopDataset(train_data_dir + "/train.csv")
+        test_dataset = TopDataset(test_data_dir + "/test.csv")
+        return train_dataset, test_dataset
+
+    def log_params(self, client, run_id):
+        client.log_param(
+            run_id=run_id, key=f"learning_rate {self._experiment_name}", value=self._lr
+        )
+        client.log_param(
+            run_id=run_id, key=f"epochs {self._experiment_name}", value=self._epochs
+        )
+        client.log_param(
+            run_id=run_id,
+            key=f"batch_size {self._experiment_name}",
+            value=self._batch_size,
+        )
+        client.log_param(
+            run_id=run_id,
+            key=f"loss {self._experiment_name}",
+            value=self.loss_.__class__.__name__,
+        )
+        client.log_param(
+            run_id=run_id,
+            key=f"optimizer {self._experiment_name}",
+            value=self.optimizer_.__class__.__name__,
+        )
+
+    def log_metrics(self, client, run_id, key, value, pipeline_level=False):
+        if pipeline_level:
+            client.log_metric(
+                run_id=run_id,
+                key=f"{self._experiment_name}/{key}",
+                value=value,
+            )
+        else:
+            client.log_metric(
+                run_id=run_id,
+                key=f"iteration_{self._iteration_num}/{self._experiment_name}/{key}",
+                value=value,
+            )
+
+    def local_train(self, checkpoint):
+        """Perform local training for a given number of epochs
+
+        Args:
+            checkpoint: Previous model checkpoint from where training has to be started.
+        """
+
+        if checkpoint:
+            self.model_.load_state_dict(torch.load(checkpoint + "/model.pt"))
+
+        with mlflow.start_run() as mlflow_run:
+            # get Mlflow client and root run id
+            mlflow_client = mlflow.tracking.client.MlflowClient()
+            root_run_id = os.environ.get("AZUREML_ROOT_RUN_ID")
```

Comment:
can we replace with mlflow_run.tags as in other scripts?

Reply:
replied in the Teams